### PR TITLE
Feature/benchmark oracle

### PR DESCRIPTION
**Pull request by gmaretic**
These changes are needed for Oracle benchmarks

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/1360)

### CODE REVIEW

**Comment by gsheasby** [comment]
A few initial questions:
* How important is this to merge today?
* What was the motivation for introducing an interface behind the enum KVSType?

(Also, seeing an interface with "Interface" in the name makes me twitch, but I don't have enough context to immediately suggest a better pair of names...)

**Comment by gmaretic** [comment]
It's not terribly urgent, although it does need to be merged in for the Oracle benchmark to be able to pick up the changes and work.

The motivation is to make it extensible, so the Oracle benchmark can just add its settings to KVSType and then execute the Atlas benchmark code.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/CassandraKeyValueServiceInstrumentation.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.performance.backend;
+
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/CassandraKeyValueServiceInstrumentation.java]: nit: double blank line

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/CassandraKeyValueServiceInstrumentation.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+    }
+
+    @Override
+    public boolean canConnect(InetSocketAddress addr) {
+        try {
+            CassandraKeyValueService.create(
+                    CassandraKeyValueServiceConfigManager.createSimpleManager(
+                            (CassandraKeyValueServiceConfig) getKeyValueServiceConfig(addr)),
+                    Optional.of(ImmutableLeaderConfig
+                            .builder()
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/CassandraKeyValueServiceInstrumentation.java]: this is inconsistent with line 39 above - I'd lean towards having `.builder` on the same line.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/KeyValueServiceTypeInterface.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+import java.net.InetSocketAddress;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+
+import com.palantir.atlasdb.keyvalue.cassandra.CassandraKeyValueService;
+import com.palantir.atlasdb.spi.KeyValueServiceConfig;
+
+public abstract class KeyValueServiceTypeInterface {
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/KeyValueServiceTypeInterface.java]: We forgot to rename this...

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/CassandraKeyValueServiceInstrumentation.java [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+                            .quorumSize(1)
+                            .localServer(addr.getHostString())
+                            .leaders(ImmutableSet.of(addr.getHostString()))
+                            .build()));
+            return true;
+        } catch (Exception e) {
+            return false;
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/CassandraKeyValueServiceInstrumentation.java]: We should probably log the exception for debugging purposes.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DatabasesContainer.java [hunk 5]
```diff
@@ -23,15 +23,15 @@
...
         try {
-            for (KeyValueServiceType backend : backends) {
+            for (KeyValueServiceTypeInterface backend : backends) {
                 DockerizedDatabase db = DockerizedDatabase.start(backend);
                 Awaitility.await()
-                        .atMost(Duration.ONE_MINUTE)
-                        .pollInterval(Duration.ONE_SECOND)
-                        .until(() -> KeyValueServiceType.canConnect(backend, db.getUri().getAddress()));
+                        .atMost(Duration.FIVE_MINUTES)
+                        .pollInterval(Duration.ONE_MINUTE)
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DatabasesContainer.java]: This seems like a rather long time to wait...

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DockerizedDatabaseUri.java [hunk 6]
```diff
@@ -33,11 +33,11 @@ public static DockerizedDatabaseUri fromUriString(String uri) {
         String[] parts = uri.trim().split(DELIMITER);
         String[] addrParts = parts[1].split(":");
         return new DockerizedDatabaseUri(
-                KeyValueServiceType.valueOf(parts[0]),
+                KeyValueServiceTypeInterface.keyValueServiceTypeFor(parts[0]),
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DockerizedDatabaseUri.java]: could rename to `KeyValueServiceInstrumentation.forDatabase`.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DockerizedDatabaseUri.java [hunk 7]
```diff
@@ -33,11 +33,11 @@ public static DockerizedDatabaseUri fromUriString(String uri) {
...
                 InetSocketAddress.createUnresolved(addrParts[0], Integer.parseInt(addrParts[1])));
     }
 
-    public KeyValueServiceType getKeyValueServiceType() {
+    public KeyValueServiceTypeInterface getKeyValueServiceType() {
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DockerizedDatabaseUri.java]: Getter should match type name, or else describe what `type` is to the `DockerizedDatabaseUri`.

Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DockerizedDatabaseUri.java]: Also, what else is this used for? If it's only used [here](https://github.com/palantir/atlasdb/pull/1360/files#diff-3f3d5b40164a1b372e79f78242e05d7eR47), then we should simply supply a `getKeyValueServiceConfig` method.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/KeyValueServiceTypeInterface.java [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+    public String getDockerComposeResourceFileName() {
+        return dockerComposeFileName;
+    }
+
+    public int getKeyValueServicePort() {
+        return kvsPort;
+    }
+
+    abstract KeyValueServiceConfig getKeyValueServiceConfig(InetSocketAddress addr);
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/KeyValueServiceTypeInterface.java]: Consider making these methods public, especially if the IDE doesn't complain at you for it.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/KeyValueServiceTypeInterface.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    abstract boolean canConnect(InetSocketAddress addr);
+
+    private static Map<String, KeyValueServiceTypeInterface> backendMap =
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/KeyValueServiceTypeInterface.java]: (optional) Consider splitting these static calls out into a factory.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/PostgresKeyValueServiceInstrumentation.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+ * http://opensource.org/licenses/BSD-3-Clause
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.palantir.atlasdb.performance.backend;
+
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/PostgresKeyValueServiceInstrumentation.java]: double blank line

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/PostgresKeyValueServiceInstrumentation.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+                .ddl(ImmutablePostgresDdlConfig.builder().build())
+                .connection(
+                        ImmutablePostgresConnectionConfig.builder()
+                                .host(addr.getHostString())
+                                .port(5432)
+                                .dbName("atlas")
+                                .dbLogin("palantir")
+                                .dbPassword(ImmutableMaskedValue.of("palantir"))
+                                .build()
+                ).build();
```
Comment (gsheasby) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/PostgresKeyValueServiceInstrumentation.java]: This builder of builders thing is a little janky - for readability, consider extracting a private getter for the connection config.

File: atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DatabasesContainer.java (hunk 5 above)
Comment (gmaretic) [inline atlasdb-perf/src/main/java/com/palantir/atlasdb/performance/backend/DatabasesContainer.java]: ONE_MINUTE was too short for Oracle, and the five minute wait (which happens only if the connection fails) is short compared to a successful run of the benchmarks.

**Comment by gmaretic** [comment]
All fixed now. Should be ready for merging.

**Pull request merged**